server/tests/Icepeak/Server/MultiSubscriptionSpec.hs

```diff
@@ -0,0 +1,340 @@
+expectNoMessage conn = do
+  Left () <- Async.race
+    (Concurrent.threadDelay 10_000) -- 0.01s
+    (WS.receiveDataMessage conn >>=
+      (\p -> expectationFailure
+        $ "Client received an unexpected payload: " <> show p))
+  pure ()
+
+withResponseJson :: WS.Connection -> (Aeson.Value -> Expectation) -> Expectation
+withResponseJson conn jsonCheck = do
+  (WS.Text payload _) <- expectDataMessage conn
+  (Just json) <- pure $ Aeson.decode payload
+  jsonCheck json
+
+invalidPayloadsSpec :: SpecWith a
+invalidPayloadsSpec = describe "Opening and sending invalid payloads" $ do
+  it "should close connection upon invalid payload" $ const $ do
+    let openThenSend dataMessage = openReusableIcepeakConn $ \conn ->
+          do WS.sendDataMessage conn dataMessage
+             Exception.catch
+               (do unexpectedDataMessage <- WS.receiveDataMessage conn
+                   pure (Unexpected $ "Received data message when socket close was expected: " <> show unexpectedDataMessage))
+               
+               (\case (WS.CloseRequest code _) -> pure $ CloseCode code
+                      otherException -> pure $ Unexpected ("Unexpected exception: " <> show otherException))
+
+    notJSON <- openThenSend (WS.Text "Hello" Nothing)
+    notJSON `shouldBe` CloseCode (Icepeak.closeCode Icepeak.TypeJsonDecodeError)
+
+    binaryMessage <- openThenSend (WS.Binary "Hello")
+    binaryMessage `shouldBe` CloseCode (Icepeak.closeCode Icepeak.TypeBinaryPayload)
+
+    notAnObject <- openThenSend (WS.Text (Aeson.encode $ Aeson.Number 3) Nothing)
+    notAnObject `shouldBe` CloseCode (Icepeak.closeCode Icepeak.TypePayloadNotObject)
+
+    unexpectedType <- openThenSend (WS.Text (Aeson.encode $ Aeson.object [ "type" .= ("subskribe" :: String) ]) Nothing)
+    unexpectedType `shouldBe` CloseCode (Icepeak.closeCode Icepeak.TypeMissingOrUnexpectedType)
+
+    noType <- openThenSend (WS.Text (Aeson.encode $ Aeson.object [ ]) Nothing)
+    noType `shouldBe` CloseCode (Icepeak.closeCode Icepeak.TypeMissingOrUnexpectedType)
+
+    outOfBounds <- openThenSend (WS.Text (BS.Lazy.replicate (fromInteger (toInteger $ Icepeak.maxPayloadBytes + 1)) 0) Nothing)
+    outOfBounds `shouldBe` CloseCode (Icepeak.closeCode Icepeak.TypeSizeOutOfBounds)
+
+    insideOfBounds <- openThenSend (WS.Text (BS.Lazy.replicate (fromInteger (toInteger Icepeak.maxPayloadBytes)) 0) Nothing)
+    insideOfBounds `shouldNotBe` CloseCode (Icepeak.closeCode Icepeak.TypeSizeOutOfBounds)
+
+singleConnectionCommunicationSpec :: SpecWith Icepeak
+singleConnectionCommunicationSpec = aroundAllWith
+  (\specUsingArgs icepeak -> openReusableIcepeakConn (curry specUsingArgs icepeak))
+  $ describe "Communication over a single connection" $ do
+  succesfulSubscribe
+  succesfulReceiveUpdates
+  succesfulUnsubscribe
+  succesfulUnsubscribeNoUpdates
+  
+succesfulSubscribe :: SpecWith (Icepeak, WS.Connection)
+succesfulSubscribe = it "should subscribe and receive success response with values"
```

Comment:
Same here, try to keep individual tests small, so there isn't any ambiguity if
we see a failing test, which assertion that test corresponds to.